### PR TITLE
Annotation autocomplete

### PR DESCRIPTION
**Pull request by robert-tan**
## Motivation and Context

https://github.com/MarkUsProject/Markus/issues/2561

## Your Changes

**Description**:
Added a dynamic dropdown that appears in the Create Annotation modal if the annotation text being typed matches existing annotations.

**Type of change** (select all that apply):

 
- [x] New feature (non-breaking change which adds functionality)

## Testing

Fixed tests in annotation_categories_controller_spec.rb

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. 
- [x] I have verified that the TravisCI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have added tests for my changes. 
- [x] I have updated the Changelog.md file. 
- [x] I have described any required documentation changes below.

### CODE REVIEW

File: app/controllers/annotation_categories_controller.rb [hunk 1]
```diff
@@ -135,14 +135,15 @@ def update_annotation_text
...
-    annotation_texts = texts_for_current_assignment.where("content LIKE ?", "#{string}%")
-    if annotation_texts.size == 1
-      render json: "#{annotation_texts.first.content}".html_safe
-    else
-      render json: ''.html_safe
-    end
+    one_time_texts = AnnotationText.joins(annotations: { result: { grouping: :group } })
+                                   .where(creator_id: params[:id], 'groupings.assessment_id': @assignment.id)
+
+    annotation_texts = texts_for_current_assignment.where("lower(content) LIKE ?", "#{string.downcase}%").limit(10) |
```
[the remark this answers is not in the document]
Comment (robert-tan) [reply app/controllers/annotation_categories_controller.rb:144]: ok mr hound

File: app/views/results/common/_annotations.js.erb [hunk 2]
```diff
@@ -298,9 +298,56 @@
...
+          $('#annotation_text_id').val(elem.id);
+          $('#new_annotation_category').val(elem.annotation_category_id);
+        }
+        annotation.innerText = elem.content;
+        annotationTextList.appendChild(annotation);
+      });
+    }).catch((e) => console.log(e));
+  }
+
+  var typing_timer;
```
Comment (david-yz-liu) [inline app/views/results/common/_annotations.js.erb:345]: Hmm, I wonder if you can put this inside `open_annotation_modal`?

File: app/views/results/common/_annotations.js.erb [hunk 3]
```diff
@@ -298,9 +298,56 @@
...
+      }
+    }).then((res) => {
+      if (res.length === 0) {
+        $('#annotation_completion').hide();
+        return;
+      }
+
+      $('#annotation_completion').show();
+      annotationTextList.innerHTML = '';
+      $('#annotation_completion_text')[0].innerText = `Did you mean? (${res.length} similar annotations found)`;
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/views/results/common/_annotations.js.erb:329]: Also, please internationalize the text. You can also use the "count" option to differentiate between 1 vs. many in the localized text. See https://github.com/fnando/i18n-js, search for "pluralization".

File: app/views/results/common/_annotations.js.erb [hunk 4]
```diff
@@ -298,9 +298,56 @@
...
+      return;
+    }
+
+    $.ajax({
+      method: "get",
+      url: Routes.find_annotation_text_assignment_annotation_categories_url(<%= assignment_id %>),
+      dataType: "json",
+      data: {
+        string: value,
+        assignment_id: <%= assignment_id %>,
```
Comment (david-yz-liu) [inline app/views/results/common/_annotations.js.erb:318]: Hmm, I'm not sure you need to include this if the assignment id is passed to the `Routes` url function above.

File: app/views/annotations/_annotation_modal.html.erb [hunk 5]
```diff
@@ -14,6 +14,18 @@
                         placeholder: t('results.annotation.placeholder'),
                         rows: 8 %>
       </p>
+      <ul class='tags' key='annotation_completion' id='annotation_completion' style='display: none'>
+        <%= hidden_field_tag 'annotation_text_id', value: '' %>
+        <li class='annotation_category'
+            id='annotation_completion_li'>
+          <p id='annotation_completion_text'> Did you mean? (0 similar annotations found)</p>
```
Comment (david-yz-liu) [inline app/views/annotations/_annotation_modal.html.erb:21]: This string should be internationalized as well.

**Pull request merged**